### PR TITLE
feat: added correctness test

### PR DESCRIPTION
**Pull request by 3benbox**

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -6,7 +6,7 @@ WORKDIR /home/builder/rust-ceramic
 # Use the same ids as the parent docker image by default
 ARG UID=1001
 ARG GID=1001
```
Comment (nathanielc) [inline Dockerfile:8]: I'll go remove the UID and GID bits from the parent docker container so we just use root. Github actions do not like custom users :(

File: .github/workflows/rust-build-and-test.yml [hunk 2]
```diff
@@ -99,5 +108,99 @@ jobs:
...
+          RELEASE_TAG=$(cargo metadata --format-version=1 --no-deps | jq '.packages[0].version' | tr -d '"')
+          # Use the release tag to deploy, if one is available.
+          BUILD_TAG=$RELEASE_TAG
+        fi
+        if [[ "${{ github.event_name }}" == "pull_request" && "${{ github.base_ref }}" == "main" ]]; then
+          # Use PR commit sha on PRs for main
+          RELEASE_TAG=$(echo ${{ github.event.pull_request.head.sha }} | head -c 12)
+          BUILD_TAG=$RELEASE_TAG
+        fi
+        make SHA="${{ github.SHA }}" SHA_TAG="$SHA_TAG" RELEASE_TAG="$RELEASE_TAG" publish-docker
```
Comment (smrz2001) [inline .github/workflows/rust-build-and-test.yml:140]: I think we can simplify this section:
```suggestion
        BUILD_TAG=$(echo ${{ github.event.pull_request.head.sha }} | head -c 12)
        make SHA_TAG="$BUILD_TAG" publish-docker
```

Comment (smrz2001) [reply .github/workflows/rust-build-and-test.yml:140]: Another thing to note is that `make publish-docker` will override `latest`. Could update `publish.sh` to take additional tags as input and pass `latest` when called from CD and nothing when called from here.

File: .github/workflows/rust-build-and-test.yml [hunk 3]
```diff
@@ -99,5 +108,99 @@ jobs:
...
+            TESTS="."
+          elif [ "${{ github.event_name}}" == "workflow_dispatch" ]; then
+            TESTS="${{ github.event.inputs.test_selector || '.' }}"
+          fi
+          ./bin/hermetic-driver test \
+            --network basic-rust.yaml \
+            --flavor correctness \
+            --test-selector ${TESTS}
 
+      - name: Clean up correctness-test
```
Comment (smrz2001) [inline .github/workflows/rust-build-and-test.yml:204]: Can remove this section?

File: .github/workflows/rust-build-and-test.yml [hunk 4]
```diff
@@ -99,5 +108,99 @@ jobs:
...
+      -
+        name: Get Network template and update
+        env:
+          RUST_CERAMIC_IMAGE_TAG: ${{ needs.build.outputs.build_tag }}
+          RUST_CERAMIC_IMAGE_REPO: public.ecr.aws/r5b3e0r5/3box/ceramic-one
+        run: |
+          set -ex
+          export RUST_CERAMIC_IMAGE=${RUST_CERAMIC_IMAGE_REPO}:${RUST_CERAMIC_IMAGE_TAG}
+          export KERAMIK_NETWORK_NAME=hermetic-rust-ceramic-${RUST_CERAMIC_IMAGE_TAG}
+          echo "KERAMIK_NETWORK_NAME=${KERAMIK_NETWORK_NAME}" >> "$GITHUB_ENV"
```
Comment (smrz2001) [inline .github/workflows/rust-build-and-test.yml:173]: Does this need to be in the GitHub env?

File: Dockerfile (hunk 1 above)
Comment (3benbox) [reply Dockerfile]: removed!

File: .github/workflows/rust-build-and-test.yml [hunk 5]
```diff
@@ -99,5 +108,99 @@ jobs:
...
+          BUILD_TAG=$RELEASE_TAG
+        fi
+        make SHA="${{ github.SHA }}" SHA_TAG="$SHA_TAG" RELEASE_TAG="$RELEASE_TAG" publish-docker
+        echo "Build tag:"
+        echo ${BUILD_TAG}
+        echo "build_tag=${BUILD_TAG}" >> $GITHUB_OUTPUT
+
+  correctness-tests:
+    runs-on: ubuntu-latest
+    environment: github-tests-2024
```
[the remark this answers is not in the document]
Comment (3benbox) [reply .github/workflows/rust-build-and-test.yml:147]: I like having the cluster name so you know where it's going to deploy to

File: .github/workflows/rust-build-and-test.yml [hunk 6]
```diff
@@ -99,5 +108,99 @@ jobs:
...
+        env:
+          RUST_CERAMIC_IMAGE_TAG: ${{ needs.build.outputs.build_tag }}
+          RUST_CERAMIC_IMAGE_REPO: public.ecr.aws/r5b3e0r5/3box/ceramic-one
+        run: |
+          set -ex
+          export RUST_CERAMIC_IMAGE=${RUST_CERAMIC_IMAGE_REPO}:${RUST_CERAMIC_IMAGE_TAG}
+          export KERAMIK_NETWORK_NAME=hermetic-rust-ceramic-${RUST_CERAMIC_IMAGE_TAG}
+          echo "KERAMIK_NETWORK_NAME=${KERAMIK_NETWORK_NAME}" >> "$GITHUB_ENV"
+          mkdir ./bin
+          sudo apt-get update && sudo apt-get install -y curl
```
[the remark this answers is not in the document]
Comment (3benbox) [reply .github/workflows/rust-build-and-test.yml:175]: removed

File: .github/workflows/rust-build-and-test.yml [hunk 7]
```diff
@@ -99,5 +108,99 @@ jobs:
...
+          curl -L https://github.com/3box/ceramic-tests/releases/download/v0.2.0/hermetic-driver-x86_64-linux -o ./bin/hermetic-driver
+          chmod +x ./bin/hermetic-driver
+          curl -LO https://raw.githubusercontent.com/3box/ceramic-tests/main/networks/basic-rust.yaml
+          ./bin/yq -i '.spec.ceramic[0].ipfs.rust.image = strenv(RUST_CERAMIC_IMAGE)' basic-rust.yaml
+          ./bin/yq -i '.metadata.name = strenv(KERAMIK_NETWORK_NAME)' basic-rust.yaml
+          ./bin/yq --no-colors basic-rust.yaml
 
+      -
+        name: Run correctness-tests
+        # test if the event is a pr (synchronize), run fast,
```
[the remark this answers is not in the document]
Comment (3benbox) [reply .github/workflows/rust-build-and-test.yml:187]: Cleaned it up, I think some info about what's **expected** to  happen is good

File: .github/workflows/rust-build-and-test.yml (hunk 3 above)
Comment (3benbox) [reply .github/workflows/rust-build-and-test.yml:204]: Removed

File: .github/workflows/rust-build-and-test.yml (hunk 4 above)
Comment (3benbox) [reply .github/workflows/rust-build-and-test.yml:173]: nope! removed

File: .github/workflows/rust-build-and-test.yml [hunk 8]
```diff
@@ -99,5 +108,99 @@ jobs:
...
+          cluster_name: ${{ vars.GKE_CLUSTER }}
+          location: ${{ vars.GKE_ZONE }}
+
+      -
+        name: Get Network template and update
+        env:
+          RUST_CERAMIC_IMAGE_TAG: ${{ needs.build.outputs.build_tag }}
+          RUST_CERAMIC_IMAGE_REPO: public.ecr.aws/r5b3e0r5/3box/ceramic-one
+        run: |
+          set -ex
```
[the remark this answers is not in the document]
Comment (3benbox) [reply .github/workflows/rust-build-and-test.yml:170]: Moved!

File: ci-scripts/get-update-network.sh [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+export KERAMIK_NETWORK_NAME=hermetic-rust-ceramic-${RUST_CERAMIC_IMAGE_TAG}
+mkdir ./bin
+curl -L https://github.com/mikefarah/yq/releases/download/v4.40.5/yq_linux_amd64 -o ./bin/yq
+chmod +x ./bin/yq
+curl -L https://github.com/3box/ceramic-tests/releases/download/v0.2.0/hermetic-driver-x86_64-linux -o ./bin/hermetic-driver
+chmod +x ./bin/hermetic-driver
+curl -LO https://raw.githubusercontent.com/3box/ceramic-tests/main/networks/basic-rust.yaml
+./bin/yq -i '.spec.ceramic[0].ipfs.rust.image = strenv(RUST_CERAMIC_IMAGE)' basic-rust.yaml
+./bin/yq -i '.metadata.name = strenv(KERAMIK_NETWORK_NAME)' basic-rust.yaml
+./bin/yq --no-colors basic-rust.yaml
```
Comment (smrz2001) [inline ci-scripts/get-update-network.sh:13]: minor: new line 😅

File: .github/workflows/rust-build-and-test.yml (hunk 8 above)
[the remark this answers is not in the document]
Comment (smrz2001) [reply .github/workflows/rust-build-and-test.yml:158]: Nice! Should we add a makefile directive as well to call from here? Can take the hermetic driver release number, the network name to retrieve from the tests repo, and the rust-ceramic image tag as inputs (similar to how it's done in the tests repo for some directives).

But you've put a ton of work into the PR already, so it's totally cool to open a ticket for this ^ and wind up this PR. I can also take care of that last bit while you're out.

**Review by smrz2001 — APPROVED** [review]
🚀 🚀 🚀

**Pull request merged**